Add resources to be able to deploy dashboards

dannyhorodniczy opened the pull request:
Added resources to able to deploy dashboards, currently the options are limited because Microsoft does not publish a template for all possible dashboard parts.

Commit:
Add resources to be able to deploy dashboards

dilmanous commented on src/PoshArmDeployment/Public/ResourceProviders/Microsoft.Portal/New-ArmDashboardsApplicationMap.ps1:
```diff
@@ -0,0 +1,50 @@
+    $ApplicationInsightsResourceName = $ApplicationInsights.Name
+    $ApplicationMap = [PSCustomObject][ordered]@{
+      PSTypeName = "DashboardPart"
+      position   = @{ }
```
Don't we care about the position and if not is including it mandatory?

scbergeron replied:
You need to call Set-ArmDashboardsPartsPosition after the creation to set it correctly

dilmanous commented on src/PoshArmDeployment/Public/ResourceProviders/Microsoft.Portal/New-ArmDashboardsApplicationMap.ps1:
```diff
@@ -0,0 +1,50 @@
+            value = @{
+              durationMs            = 86400000
+              endTime               = $null
+              createdTime           = '2018-05-04T01:20:33.345Z'
```
What does this time do and why is it hard-coded?

scbergeron replied:
The time is not doing anything but i'll replace it by a function call

dilmanous commented on src/PoshArmDeployment/Public/ResourceProviders/Microsoft.Portal/New-ArmDashboardsCuratedBladeFailuresPinned.ps1:
```diff
@@ -0,0 +1,46 @@
+              timeContext = @{
+                durationMs            = 86400000
+                endTime               = $null
+                createdTime           = '2018-05-04T01:20:33.345Z'
```
why do we hard code this time?

mytev commented on src/PoshArmDeployment/Public/ResourceProviders/Microsoft.Portal/Add-ArmDashboardsPartsElement.ps1:
```diff
@@ -0,0 +1,17 @@
+  [CmdletBinding(SupportsShouldProcess = $true)]
+  [OutputType("Dashboards")]
+  Param(
+    [PSTypeName("DashboardPart")]
```
`[Parameter(Mandatory, ValueFromPipeline)]` ?

mytev commented on src/PoshArmDeployment/Public/ResourceProviders/Microsoft.Portal/Add-ArmDashboardsPartsElement.ps1:
```diff
@@ -0,0 +1,17 @@
+  )
+  If ($PSCmdlet.ShouldProcess("Adding DashboardPart to Dashboards")) {
+    $index = ($Dashboards.properties.lenses.'0'.parts | Get-Member -View Extended).count 
+    $Dashboards.properties.lenses.'0'.parts | Add-Member -MemberType NoteProperty -Name $index.ToString() -Value $Part
```
Why are we using $index.ToString() for the name?

mytev commented on src/PoshArmDeployment/Public/ResourceProviders/Microsoft.Portal/New-ArmDashboardsApplicationMap.ps1:
```diff
@@ -0,0 +1,50 @@
+  [OutputType("DashboardPart")]
+  Param(
+    [PSTypeName("ApplicationInsights")]
+    [Parameter(Mandatory)]
```
`[Parameter(Mandatory, ValueFromPipeline)]` ?

mytev commented on src/PoshArmDeployment/Public/ResourceProviders/Microsoft.Portal/New-ArmDashboardsAspNetOverviewPinned.ps1:
```diff
@@ -0,0 +1,33 @@
+        }
+        defaultMenuItemId = 'overview'
+      }      
+    }
```
Should we add ApplicationInsights as a dependency in `dependsOn`?

mytev commented on src/PoshArmDeployment/Public/ResourceProviders/Microsoft.Portal/New-ArmDashboardsCuratedBladeFailuresPinned.ps1:
```diff
@@ -0,0 +1,46 @@
+  [OutputType("DashboardPart")]
+  Param(
+    [PSTypeName("ApplicationInsights")]
+    [Parameter(Mandatory)]
```
`[Parameter(Mandatory, ValueFromPipeline)]`

mytev commented on src/PoshArmDeployment/Public/ResourceProviders/Microsoft.Portal/New-ArmDashboardsCuratedBladeFailuresPinned.ps1:
```diff
@@ -0,0 +1,46 @@
+    $ApplicationInsightsResourceId = $ApplicationInsights._ResourceId
+    $CuratedBlade = [PSCustomObject][ordered]@{
+      PSTypeName = "DashboardPart"
+      position   = @{ }
```
Extra space?

mytev commented on src/PoshArmDeployment/Public/ResourceProviders/Microsoft.Portal/New-ArmDashboardsCuratedBladePerformancePinned.ps1:
```diff
@@ -0,0 +1,46 @@
+  [OutputType("DashboardPart")]
+  Param(
+    [PSTypeName("ApplicationInsights")]
+    [Parameter(Mandatory)]
```
`[Parameter(Mandatory, ValueFromPipeline)]`

dilmanous commented on src/PoshArmDeployment/Public/ResourceProviders/Microsoft.Portal/New-ArmDashboardsMonitorChart.ps1:
```diff
@@ -0,0 +1,39 @@
+  If ($PSCmdlet.ShouldProcess("Adding MonitorChartPart to Dashboards")) {
+    $MonitorChart = [PSCustomObject][ordered]@{
+      PSTypeName = "DashboardPart"
+      position   = @{ }
```
why not receive position as a function parameter with default value of empty?

scbergeron replied:
I thought it made more sense to create the element then to decide it's position. The creation already has alot of parameter. Default values would not be very useful because you always want to decide the position and the size of the element.

dilmanous commented on src/PoshArmDeployment/Public/ResourceProviders/Microsoft.Portal/New-ArmDashboardsUsageUsersOverview.ps1:
```diff
@@ -0,0 +1,50 @@
+              createdTime           = '2018-05-04T01:20:33.345Z'
+              isInitialTime         = $false
+              grain                 = 1
+              useDashboardTimeRange = $false
```
why hardcoded?

dilmanous commented on src/Tests/Public/ResourceProviders/Microsoft.Portal/Add-ArmDashboardsPartsElement.Tests.ps1:
```diff
@@ -0,0 +1,209 @@
+                        }
+                        resources   = @()
+                        dependsOn   = @()
+                    }; Part  = [PSCustomObject][ordered]@{
```
put on separate line?

dilmanous commented on src/Tests/Public/ResourceProviders/Microsoft.Portal/New-ArmDashboardsAnalyticsChart.Tests.ps1:
```diff
@@ -0,0 +1,105 @@
+    | New-ArmApplicationInsightsResource -Location $ExpectedLocation
+
+    Context "Unit tests" {
+      It "Given a '<Source>', '<SubscriptionId>', '<ResourceGroupName>', '<Query>', '<Dimensions>', '<ResourceType>', '<ChartType>', '<Title>', '<SubTitle>' it returns '<Expected>'" -TestCases @(
```
break into two lines for readability?

dilmanous commented on src/Tests/Public/ResourceProviders/Microsoft.Portal/New-ArmDashboardsApplicationMap.Tests.ps1:
```diff
@@ -0,0 +1,57 @@
+InModuleScope PoshArmDeployment {
+  Describe "New-ArmDashboardsApplicationMap" {
+    $InsightsResourceType = "microsoft.insights/components"
+    $ExpectedLocation = "SomeLocation"
```
unnecessary variable

dilmanous commented on src/Tests/Public/ResourceProviders/Microsoft.Portal/New-ArmDashboardsAspNetOverviewPinned.Tests.ps1:
```diff
@@ -0,0 +1,44 @@
+InModuleScope PoshArmDeployment {
+  Describe "New-ArmDashboardsAspNetOverviewPinned" {
+    $InsightsResourceType = "microsoft.insights/components"
+    $ExpectedLocation = "SomeLocation"
```
not necessary

dilmanous commented on src/Tests/Public/ResourceProviders/Microsoft.Portal/New-ArmDashboardsCuratedBladeFailuresPinned.Tests.ps1:
```diff
@@ -0,0 +1,57 @@
+InModuleScope PoshArmDeployment {
+  Describe "New-ArmDashboardsCuratedBladeFailuresPinned" {
+    $InsightsResourceType = "microsoft.insights/components"
+    $ExpectedLocation = "SomeLocation"
```
not necessary

dilmanous commented on src/Tests/Public/ResourceProviders/Microsoft.Portal/New-ArmDashboardsCuratedBladePerformancePinned.Tests.ps1:
```diff
@@ -0,0 +1,57 @@
+InModuleScope PoshArmDeployment {
+  Describe "New-ArmDashboardsCuratedBladePerformancePinned" {
+    $InsightsResourceType = "microsoft.insights/components"
+    $ExpectedLocation = "SomeLocation"
```
not necessary

dilmanous commented on src/Tests/Public/ResourceProviders/Microsoft.Portal/Set-ArmDashboardsPartsPosition.Tests.ps1:
```diff
@@ -0,0 +1,84 @@
+          -ColSpan $ColSpan `
+          -RowSpan $RowSpan
+
+        ($actual | ConvertTo-Json -Compress | ForEach-Object { [System.Text.RegularExpressions.Regex]::Unescape($_) }) `
```
specify the depth of assertion?

scbergeron replied:
No because the depth of 1 is fine

mytev commented on src/PoshArmDeployment/Public/ResourceProviders/Microsoft.Portal/New-ArmDashboardsCuratedBladePerformancePinned.ps1:
```diff
@@ -0,0 +1,46 @@
+            value      = @{
+              version     = '1.0.0'
+              timeContext = @{
+                durationMs            = 86400000
```
Should we move these values in the parameters?

mytev commented on src/PoshArmDeployment/Public/ResourceProviders/Microsoft.Portal/New-ArmDashboardsQuickPulseButtonSmall.ps1:
```diff
@@ -0,0 +1,43 @@
+      PSTypeName = "DashboardPart"
+      position   = @{ }
+      metadata   = @{
+        inputs = @(@{
```
It seems that this code is reused at a lot of places 
```
@{
  name  = 'ComponentId'
  value = @{
    Name = $ApplicationInsightsResourceName
    SubscriptionId = $SubscriptionId
    ResourceGroup  = $ResourceGroupName
  }
}
```

Should we extract it?

mytev commented on src/PoshArmDeployment/Public/ResourceProviders/Microsoft.Portal/New-ArmDashboardsWebTestsPinned.ps1:
```diff
@@ -0,0 +1,62 @@
+              ResourceGroup  = $ResourceGroupName
+            }
+          }, 
+          @{
```
The TimeContext is also reused at a lot of place
```
          @{
            name  = 'TimeContext'
            value = @{
              durationMs            = 86400000
              endTime               = $null
              createdTime           = '2018-05-04T01:20:33.345Z'
              isInitialTime         = $true
              grain                 = 1
              useDashboardTimeRange = $false
            }
          }
```
Is it possible to extract it?

Commit:
Add missing tests
Add method to create chart dimension, chart visualization and chart metrics
Fixed comments

Commit:
Fix comments (missing tests)

Commit:
Add test

Commit:
Add BladeOnClick tests

Commit:
Add missing tests

Commit:
Merge branch 'feature/dashboard-deployment' of https://github.com/jackdurkalik/PoshArm-Deployment into feature/dashboard-deployment

Commit:
Add and fix tests

Commit:
Fix tests

scbergeron committed:
Merge pull request #1 from jackdurkalik/mster

Fix tests

pcherpitel committed:
Update private DNS zone to match new resource schema (#37)

* Add private dns zones

Co-authored-by: danieljosephh <horodniczy.danny@gmail.com>

pcherpitel committed:
Add support for static metric alert criteria and update action group to support email receivers and arm role receivers. (#38)

Commit:
Merge branch 'master' into feature/dashboard-deployment
